src/transaction/transaction_group_complete_manager.cpp

```diff
@@ -39,55 +46,165 @@ namespace cubtx
+	m_group_complete_condvar.wait (ulock, [&] {return is_group_logged (group_id);});
+
+	er_log_group_complete_debug (ARG_FILE_LINE, "group_complete_manager::complete_logging: (%llu)\n", group_id);
+	return;
+      }
+#endif
+
+    /* I'm the only thread. All completes are done by me. */
+    cubthread::entry *thread_p = &cubthread::get_entry ();
+    do_prepare_complete (thread_p);
+    do_complete (thread_p);
+    assert (is_group_logged (group_id));
+    er_log_group_complete_debug (ARG_FILE_LINE, "group_complete_manager::complete_logging: (%llu)\n", group_id);
   }
 
-  //
-  // wait_for_complete waits for complete event on specified group_id.
-  //
-  void group_complete_manager::wait_for_complete (id_type group_id)
+  void group_complete_manager::complete (id_type group_id)
   {
     if (group_id < m_latest_closed_group_id)
       {
 	/* Already advanced to next group. No need to acquire mutex. */
+	er_log_group_complete_debug (ARG_FILE_LINE, "group_complete_manager::complete: (tran_index = %d, group_id = %llu)\n",
+				     LOG_FIND_THREAD_TRAN_INDEX (&cubthread::get_entry ()), group_id);
 	return;
       }
 
-    std::unique_lock<std::mutex> ulock (m_group_complete_mutex);
-    /* TODO - consider stop and optimize next call */
-    m_group_complete_condvar.wait (ulock, [&] {return is_group_completed (group_id);});
+#if defined (SERVER_MODE)
+    if (BO_IS_SERVER_RESTARTED ())
+      {
+	if (group_id == m_latest_closed_group_id)
+	  {
+	    if (is_latest_closed_group_completed ())
+	      {
+		/* Group_id complete. No need to acquire mutex. */
+		assert (group_id <= m_latest_closed_group_id);
+		er_log_group_complete_debug (ARG_FILE_LINE, "group_complete_manager::complete: (tran_index = %d, group_id = %llu)\n",
+					     LOG_FIND_THREAD_TRAN_INDEX (&cubthread::get_entry ()), group_id);
+		return;
+	      }
+	  }
+
+	if (!need_wait_for_complete ())
+	  {
+	    /* Async case. */
+	    er_log_group_complete_debug (ARG_FILE_LINE, "group_complete_manager::complete: (tran_index = %d, group_id = %llu)\n",
+					 LOG_FIND_THREAD_TRAN_INDEX (&cubthread::get_entry ()), group_id);
+	    return;
+	  }
+
+	/* Waits for complete event on specified group_id. */
+	std::unique_lock<std::mutex> ulock (m_group_complete_mutex);
+	/* TODO - consider whether is better to use stop and wait with timeout */
+	m_group_complete_condvar.wait (ulock, [&] {return is_group_completed (group_id);});
+	er_log_group_complete_debug (ARG_FILE_LINE, "group_complete_manager::complete: (tran_index = %d, group_id = %llu)\n",
+				     LOG_FIND_THREAD_TRAN_INDEX (&cubthread::get_entry ()), group_id);
+
+	return;
+      }
+#endif
+
+    /* I'm the only thread. All completes are done by me. */
+    cubthread::entry *thread_p = &cubthread::get_entry ();
+    do_prepare_complete (thread_p);
+    do_complete (thread_p);
+    assert (is_group_completed (group_id));
+    er_log_group_complete_debug (ARG_FILE_LINE, "group_complete_manager::complete: (%llu)\n", group_id);
```

Comment:
complete, complete_logging & complete_mvcc look very similar. Can't they be somehow unified?

Reply:
Only in SA mode, since same code repeats.